packages/graphql-codegen-cli/tests/watcher.spec.ts

```diff
@@ -0,0 +1,548 @@
+        // watch-doc (matched in global watch, excluded in local doc)
+        './foo/match-watch-doc-everywhere.graphql',
+        './foo/fizz/match-watch-doc-everywhere.graphql',
+        './foo/fizz/buzz/match-watch-doc-everywhere.graphql',
+        './foo/fizz/buzz/foobarbaz/match-watch-doc-everywhere.graphql',
+        // watch-schema (matched in global watch, excluded in local schema)
+        './foo/match-watch-schema-everywhere.graphql',
+        './foo/fizz/match-watch-schema-everywhere.graphql',
+        './foo/fizz/buzz/match-watch-schema-everywhere.graphql',
+        './foo/fizz/buzz/foobarbaz/match-watch-schema-everywhere.graphql',
+        // doc
+        './foo/match-doc-everywhere.graphql',
+        './foo/fizz/match-doc-everywhere.graphql',
+        './foo/fizz/buzz/match-doc-everywhere.graphql',
+        './foo/fizz/buzz/foobarbaz/match-doc-everywhere.graphql',
+        // schema
+        './foo/match-schema-everywhere.graphql',
+        './foo/fizz/match-schema-everywhere.graphql',
+        './foo/fizz/buzz/match-schema-everywhere.graphql',
+        './foo/fizz/buzz/foobarbaz/match-schema-everywhere.graphql',
+      ],
+    });
+  });
+
+  test('globally negated paths should be excluded even when a local pattern matches them', async () => {
+    const mockWatcher = await setupMockWatcher({
+      filepath: './foo/some-config.ts',
+      config: {
+        schema: ['!**/exclude-schema-everywhere.graphql'],
+        watch: [
+          '!**/exclude-watch-everywhere.graphql',
+          '!**/exclude-watch-doc-everywhere.graphql',
+          '!**/exclude-watch-schema-everywhere.graphql',
+        ],
+        documents: ['!**/exclude-doc-everywhere.graphql'],
+        generates: {
+          ['./foo/local-inclusions-dont-precede-global-exclusions.ts']: {
+            watchPattern: [
+              'foo/global-beats-local/exclude-watch-everywhere.graphql',
+              'foo/global-beats-local/exclude-watch-doc-everywhere.graphql',
+              'foo/global-beats-local/exclude-watch-schema-everywhere.graphql',
+            ],
+            documents: [
+              'foo/global-beats-local/exclude-doc-everywhere.graphql',
+              'foo/global-beats-local/exclude-watch-doc-everywhere.graphql',
+            ],
+            schema: [
+              'foo/global-beats-local/exclude-schema-everywhere.graphql',
+              'foo/global-beats-local/exclude-watch-schema-everywhere.graphql',
+            ],
+          },
+        },
+      },
+    });
+
+    expect(mockWatcher.watchDirectory).toBe(join(process.cwd(), 'foo'));
+
+    await assertBuildTriggers(mockWatcher, {
+      shouldNotTriggerBuild: [
+        'foo/global-beats-local/exclude-watch-everywhere.graphql',
+        'foo/global-beats-local/exclude-doc-everywhere.graphql',
+        'foo/global-beats-local/exclude-schema-everywhere.graphql',
+        '!**/exclude-watch-doc-everywhere.graphql',
+        '!**/exclude-watch-schema-everywhere.graphql',
```

Comment:
oops, small typo in this assertion. Fixing now, and making my final commit 😅

Reply:
fixed & force pushed (the tests still pass)

Reply:
Gah, after this 2 line change, a [GHA workflow failed](https://github.com/dotansimha/graphql-code-generator/actions/runs/4647277515/jobs/8224186329?pr=9275) ("Testing / Examples (pull_request) Failing after 9m")

I think this is a flaky test unrelated to my changes. I'm going to force push again.